### DISCUSSION TITLE
Request : Double-Byte characters need to be counted as 2

### DISCUSSION THREAD

**Discussion by TL-aalive** [SubtitleEdit/subtitleedit#5711] [Ideas]
Hello, I am from South Korea, and I'd like to thanks to this community for making such a wonderful program.

I am using SubtitleEdit to translate English to Korean.  As you may know, Korean, Japanese and Chinese character set are Double-Byte based. And it causes us 'character counting' issues. 

Since SubtitleEdit doesn't distinguish Single-Bye from Double-Byte, it counts '백' and 'B' as 1 character in counting.  
But in reality, every OTT company requires Double-Byte characters to be counted as 1 and Single-Byte characters to be counted as 0.5.

Let me show you a Guideline from Netflix for 'Character Limitation' as below;
**2. Character Limitation

- 16 characters per line
- Latin characters, spaces, punctuation count as 0.5 character**
( Source : https://partnerhelp.netflixstudios.com/hc/en-us/articles/216001127-Korean-Timed-Text-Style-Guide )

The translators in Korea actually count the characters line by line to  follow the Guideline and it is painful. 

- Suggestion : Add option to count Double-Byte characters as 2, or count Single-Byte characters as 0.5

**Comment by niksedk** [comment]
Does the spec not say the Korean letters are counted as *one*? (probably a typo in the title?)

**Reply by TL-aalive** [reply]
I made an assumption that counting ASCII as 1 and CJK as 2 is easier than counting ASCII as 0.5 and CJK as 1.  But I got confused when I wrote the idea.  Any way, as long as ASCII : CJK keeps 1 : 2 ratio, either way is fine for me.  Your current Beta CPS option looks good to me since we can sync with Netflix guideline of 16 characters per line.

**Comment by niksedk** [comment]
How does this work? https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.4/SubtitleEditBeta.zip

You need to change "cps/line length style" first:
![image](https://user-images.githubusercontent.com/1261331/150728201-a9b8b18f-ed8f-4ad7-9bb9-ebb4c9f24631.png)

ping @TL-aalive

**Reply by TL-aalive** [reply]
![option](https://user-images.githubusercontent.com/97822423/150891157-a924731e-d380-4b53-b346-fd1439e213e6.png)

The beta version is not working yet.
After set the CPS/line-length as above, the beta count drops to half of original version count as below.


![CPS1](https://user-images.githubusercontent.com/97822423/150891303-105e1c4c-1bea-4c6d-97f4-14353b40a9d2.png)

BETA counts 4.5, It should count as 8.

![CPS-original](https://user-images.githubusercontent.com/97822423/150891363-8e21e462-fb40-4153-a47b-8afa1f7d13cc.png)

Original version counts 9.

**Reply by TL-aalive** [reply]
It looks like the BETA counts all characters as 0.5

**Comment by TL-aalive** [comment]
Thanks a lot!I will get back to you once test this beta. It looks OK for me.
    2022년 1월 24일 월요일 오후 02시 42분 32초 GMT+9, Nikolaj Olsson ***@***.***> 작성:  
 
 


How does this work? https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.4/SubtitleEditBeta.zip

You need to change "cps/line length style" first:


—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by niksedk** [comment]
thx for testing :)
Beta updated to hopefully detect CJK chars better: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.4/SubtitleEditBeta.zip

**Reply by LeonCheung** [reply]
Hmm, The default SE rule is like "**Full-width 1, half-width 1**". I think what @TL-aalive wants is "**Full-width 1, half-width 0.5**", simply a way to treat half-width character as 0.5. However, one problem I noticed for now is that when using the 3rd and 4th in the list ( CJK-related), at least Chinese punctuations (full-width) are also counted as 0.5, which is unexpected. Maybe the same case in Korean and Japanese?

For example, `你好，世界。` should be counted as 6 here, but it's 5 in beta 404:

![image](https://user-images.githubusercontent.com/494788/150925605-ef1f1e8a-a31e-4c5e-ba19-75d4da3fa2fb.png)

**Reply by TL-aalive** [reply]
I cannot find beta 404.  But in Korean, punctuations are single-byte(half-width) and should be counted as 0.5. 
If you used "CJK 1, latin 0.5, space 0" option for that count, it will work for Korean.

**Reply by TL-aalive** [reply]
I found Japanese character limitation rule from Netflix as below.
So, in my opinion, Korean and Japanese could be counted in single option.
 
I.5. Character Limitation

Horizontal subtitles - Maximum 13 full-width characters per line
Vertical subtitles - Maximum 11 full-width characters per line
Full-width character, space, and punctuation counts as 1 character
Half-width character, space, and punctuation counts as 0.5 characters
(source : https://partnerhelp.netflixstudios.com/hc/en-us/articles/215767517-Japanese-Timed-Text-Style-Guide )

**Comment by niksedk** [comment]
OK, how is this: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.4/SubtitleEditBeta.zip ?

Change set: https://github.com/SubtitleEdit/subtitleedit/commit/38314167912007e0bba99948b9134273493ec45b

**Reply by TL-aalive** [reply]
This BETA 407 is same as 402. Can I get BETA 404 for test?

**Reply by niksedk** [reply]
Did you change the "line length/cps style" ?
![image](https://user-images.githubusercontent.com/1261331/150943437-0dc81616-7fb2-4820-89bc-12190feb1425.png)

**Reply by TL-aalive** [reply]
![407](https://user-images.githubusercontent.com/97822423/150945932-ceb595b6-b1ab-4a7a-8e16-ae0fc598c7ac.png)
![옵션](https://user-images.githubusercontent.com/97822423/150945940-a42da8db-b8f5-4f01-a11f-652c0e325ec4.png)
![카운트](https://user-images.githubusercontent.com/97822423/150945949-0b8814f7-8c78-4e13-bb1a-76ffe463daa4.png)

Yes, I did.

**Comment by LeonCheung** [comment]
@niksedk Ah, now it's expected. Chinese uses full-width punctuations (most of the cases). Here is a full list of full-width punctuations we use (totally 27, and the first two are just added in the last commit):

，。、：；？！…“”—‘’（）【】「」『』〔〕《》〈〉

**Reply by niksedk** [reply]
thx :)
Chars added here: https://github.com/SubtitleEdit/subtitleedit/commit/c4602b28cc8f25f016d00630d8574294cabec745
And fix Japanese: https://github.com/SubtitleEdit/subtitleedit/commit/8e3aebe4abae7e3c90e9088ca9dc43224342356c

**Comment by niksedk** [comment]
Beta updated: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.4/SubtitleEditBeta.zip

@TL-aalive: Could you add the Korean text as plain text, if it still does not work?

**Reply by TL-aalive** [reply]
Thanks a lot!  Some changes in BETA 410

- I typed, pasted and imported as plain text, the results are same 
- Some Korean characters counted as 1, but most are still counted as 0.5
- So far, I find below 24 characters are counted as 1, and all the other Korean characters counted as 0.5
가나다라마바사아자차카타파하아야어여오요우유으이

![가나다라](https://user-images.githubusercontent.com/97822423/150980896-b45ea712-785a-4f81-a38d-09a07ff24327.png)
![대한민국](https://user-images.githubusercontent.com/97822423/150980924-a5a79984-e3bc-444e-84ff-bb987bccf076.png)

**Reply by TL-aalive** [reply]
Those 24 Korean characters represent 10 vowels and  14 consonants of Korean alphabet. Hope this helps.

**Reply by niksedk** [reply]
OK, how is this beta: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.4/SubtitleEditBeta.zip ?

Could write the Korean characters that does not work here on GitHub in plain text (and not in a screenshot).

**Reply by TL-aalive** [reply]
Oh!  BETA 412 is perfect. Great work! Thank you!

![이미지 5](https://user-images.githubusercontent.com/97822423/150989532-d7a7b49a-434f-4f46-a7f7-97480b391e4e.png)
![이미지 6](https://user-images.githubusercontent.com/97822423/150989544-e9c75b31-21e2-453b-9df5-63aedcfca897.png)

* per your request, some Korean characters that didn't work in previous version

대한민국 활화산 동물들 천국의 섬 유독 춤을 춥니다

**Reply by niksedk** [reply]
Cool, and thx for testing :)

**Reply by TL-aalive** [reply]
Thanks to you!  I asked my colleagues to test the BETA for further check.  BTW when do you expect to release 3.6.5?
